### DISCUSSION TITLE
A simple benchmark of different ways to draw shapes to canvas

### DISCUSSION THREAD

**Discussion by viblo** [pyodide/pyodide#3680] [Show and tell]
Hello!

I wanted to draw some shapes to canvas for a project Im working on. Instead of guessing the performance I made a simple benchmark of different methods. I couldnt find any similar info already, post here to help others, and also get input if I missed something.  

In short, I could think of 3 (4) different methods to draw shapes to canvas:

1. Draw on the Python side using Pillow to a Pillow Image, and then copy that over to the canvas (this was my first idea).
2. Call canvas draw methods directly from Python.
3. Construct a simple json of what needs to be drawn in Python, call a JS method with it and finally draw it from JS side.
4. As a baseline to compare, I also implemented the drawing code fully on JS side.

Each method fills the canvas with a rect, draws one circle and finally draws one line. This is then repeated 250 / 500 times and finally I run the test 100 times and take the smallest time. For fun I tested the drawing methods from the 3 browsers I had available, FF & Edge on my laptop, and Chrome on my phone.

**Numbers**
(time to render a frame in ms)

Method (250) | Firefox | Edge | Chrome (mobile)
-- | -- | -- | --
pil/pillow | 6 | 7 | 8
python canvas | 28 | 20 | 27
json | 13 | 8 | 10
pure js | 3 | 2 | 2

Method (500) | Firefox | Edge | Chrome (mobile)
-- | -- | -- | --
pil/pillow | 14 | 14 | 15
python canvas | 55 | 40 | 52
json | 32 | 19 | 24
pure js | 11 | 4 | 4

I was surprised that my (new) mobile could keep up with the laptop. Also it seems like the first idea, draw fully on the Python side to a Pillow image is not at all a bad idea, it also has the upside that the code can be reused for non-web/pyodide usage which is nice!

I put it all on jsfiddle, you can try yourself here:
https://jsfiddle.net/nyqvp0w7/

**Comment by viblo** [comment]
Small addition, today I made the same thing using pygame (running native, not through pyodide/wasm), and its about half the speed of pure js.

**Comment by ryanking13** [comment]
Thanks for sharing the benchmark!
